protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,559 @@
+    }
+}
+
+/// Remove a port mapping on the gateway.
+async fn remove_port_mapping<P: Gateway + 'static>(gateway: Arc<P>, mapping: Mapping) -> Event {
+    match P::remove_port(gateway, mapping.protocol, mapping.internal_addr.port()).await {
+        Ok(()) => Event::Removed(mapping),
+        Err(err) => Event::RemovalFailure(mapping, err),
+    }
+}
+
+/// A [`Gateway`] event.
+#[derive(Debug)]
+enum Event {
+    /// Port was successfully mapped.
+    Mapped(Mapping),
+    /// There was a failure mapping port.
+    MapFailure(Mapping, Box<dyn Error + Send + Sync + 'static>),
+    /// Port was successfully removed.
+    Removed(Mapping),
+    /// There was a failure removing the mapping port.
+    RemovalFailure(Mapping, Box<dyn Error + Send + Sync + 'static>),
+}
+
+/// Mapping of a Protocol and Port on the gateway.
+#[derive(Debug, Clone)]
+struct Mapping {
+    listener_id: ListenerId,
+    protocol: Protocol,
+    multiaddr: Multiaddr,
+    internal_addr: SocketAddrV4,
+}
+
+impl Hash for Mapping {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.listener_id.hash(state);
+    }
+}
+
+impl PartialEq for Mapping {
+    fn eq(&self, other: &Self) -> bool {
+        self.listener_id == other.listener_id
+    }
+}
+
+impl Eq for Mapping {}
+
+impl Borrow<ListenerId> for Mapping {
+    fn borrow(&self) -> &ListenerId {
+        &self.listener_id
+    }
+}
+
+/// Current state of a [`Mapping`].
+enum MappingState {
+    /// Port mapping is inactive, will be requested or re-requested on the next iteration.
+    Inactive,
+    /// Port mapping/removal has been requested on the gateway.
+    Pending,
+    /// Port mapping is active with the inner timeout.
+    Active(Delay),
+    /// Port mapping is permanent on the Gateway.
+    Permanent,
+}
+
+/// Current state of the UPnP [`Gateway`].
+enum GatewayState<P: Gateway> {
+    Searching(BoxFuture<'static, Result<(P, Ipv4Addr), Box<dyn std::error::Error>>>),
+    Available((Arc<P>, Ipv4Addr)),
+    GatewayNotFound,
+}
+
+/// A `NetworkBehaviour` for UPnP port mapping. Automatically tries to map the external port
+/// to an internal address on the gateway on a `FromSwarm::NewListenAddr`.
+pub struct Behaviour<P>
+where
+    P: Gateway,
+{
+    /// Gateway config.
+    config: Config,
+    /// UPnP interface state.
+    state: GatewayState<P>,
+
+    /// List of port mappings.
+    mappings: HashMap<Mapping, MappingState>,
+
+    /// Pending gateway events.
+    events_queue: Receiver<Event>,
+
+    /// Events sender.
+    events_sender: UnboundedSender<BoxFuture<'static, Event>>,
+}
+
+impl<P> Default for Behaviour<P>
+where
+    P: Gateway + 'static,
+{
+    fn default() -> Self {
+        Self::new(Config::default())
+    }
+}
+
+impl<P> Behaviour<P>
+where
+    P: Gateway + 'static,
+{
+    /// Builds a new `UPnP` behaviour.
+    pub fn new(config: Config) -> Self {
+        #![allow(clippy::disallowed_methods)]
+        let (events_sender, mut task_receiver) = mpsc::unbounded();
```

Comment:
Hmm, why can this one not be bounded? It is unlikely to ever be full I think but we can just drop messages on the sending side in case it is full. Better than using an unbounded channel :)